src/Workspaces/Remote/ServiceHub/Services/Renamer/RemoteRenamerService.cs

```diff
@@ -38,6 +34,27 @@ private CodeCleanupOptionsProvider GetClientOptionsProvider(RemoteServiceCallbac
             => new ClientCodeCleanupOptionsProvider(
                 (callbackId, language, cancellationToken) => _callback.InvokeAsync((callback, cancellationToken) => callback.GetOptionsAsync(callbackId, language, cancellationToken), cancellationToken), callbackId);
 
+        public ValueTask KeepAliveAsync(
+            Checksum solutionChecksum,
+            RemoteServiceCallbackId callbackId,
+            CancellationToken cancellationToken)
+        {
+            // First get the solution, ensuring that it is currently pinned.
+            return RunServiceAsync(solutionChecksum, async solution =>
+            {
+                // Once we have it, let our caller know so that it can proceed to it's next steps.
+                await _callback.InvokeAsync((callback, cancellationToken) =>
+                    callback.KeepAliveAsync(callbackId, cancellationToken), cancellationToken).ConfigureAwait(false);
+
+                // Finally wait for our caller to tell us to cancel.  That way we can release this solution and allow it
+                // to be collected if not needed anymore.
+                var taskCompletionSource = new TaskCompletionSource<bool>();
+                cancellationToken.Register(() => taskCompletionSource.TrySetCanceled(cancellationToken));
+
+                await taskCompletionSource.Task.ConfigureAwait(false);
```

Comment:
this is the sneaky part.  the call from the host to OOP gets here and will block until the entire operation is finally canceled.  Because we're inside RunServiceAsync that means we're pinning 'solution' in our internal cache attached to 'solutionChecksum'.  So all successive calls to that checksum will see that solution.

Reply:
Do we need to hold onto a thread for this, or does this yield until the task completion source triggers?

Reply:
the latter.  there's technically no threads here either (which is good).  This is the idea behind a TaskCOmpletionSource, it's a way to get a task that you programatically control the completion of (either giving it a result, an exception, or cancellation).